packages/node-app/src/controllers/giveaway.controller.ts

```diff
@@ -67,6 +67,78 @@ export const createGiveaway = async (req: Request, res: Response) => {
     res.status(500).json({ error: error.message });
   } finally {
     // remove image from tmp folder
-    fs.unlink(file.path, () => { return })
+    fs.unlink(file.path, () => { return });
+  }
+};
+
+export const updateGiveaway = async (req: Request, res: Response) => {
+  const file = req.file as Express.Multer.File
+  try {
+    const { id } = req.params;
+    const { title, description, prize, rules } = req.body;
+    const image = file ? fileToBase64(file as Express.Multer.File) : undefined
+    const updateFields = getDefinedFields({ title, description, prize, rules, image });
+    const giveaway = await Giveaway.findByIdAndUpdate(id, updateFields, { new: true });
+
+    res.status(200).json(giveaway);
+  } catch (error) {
+    res.status(500).json({ error: error.message });
+  } finally {
+    if (file) fs.unlink(file.path, () => { return });
+  }
+};
+
+export const addParticipant = async (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+    const { participant } = req.body;
+    const giveaway = await Giveaway.findById(id);
+
+    if (!giveaway)
+      return res.status(404).json({ error: 'Giveaway not found' });
+
+    if (!isValidParticipant(participant, giveaway.requirements))
+      return res.status(400).json({ error: "Participant is not whitelisted" });
+
+    const address = getParticipantAddress(participant);
+    giveaway.participants.push(address);
+    await giveaway.save();
+
+    await giveawaysContract.methods
+      .addParticipant(
+        objectIdToBytes24(giveaway._id),
+        address)
+      .send({ from: process.env.OWNER_ACCOUNT_ADDRESS, gas: '1000000' });
+
+    res.status(200).json(giveaway);
+  } catch (error) {
+    res.status(500).json({ error: error.message });
+  }
+};
+
+export const generateWinners = async (req: Request, res: Response) => {
+  try {
+    const { id } = req.params;
+    const giveaway = await Giveaway.findById(id);
+
+    if (!giveaway)
+      return res.status(404).json({ error: 'Giveaway not found' });
+    
+    if (new Date() < giveaway.endTime)
+      return res.status(400).json({ error: 'Giveaway has not ended yet' });
+
+    await giveawaysContract.methods
+      .generateWinners(objectIdToBytes24(giveaway._id))
+      .send({ from: process.env.OWNER_ACCOUNT_ADDRESS, gas: '1000000' });
```

Comment:
what does the gas value represents here?

Reply:
Since this operation involves changing the blockchain we will have to pay a gas fee for that. This gas value is the max value we allow to be charged to the owner account for this transaction

Reply:
I think it is worth to add a small comment for that `// max contract owner charge`